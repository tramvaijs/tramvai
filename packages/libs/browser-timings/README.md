# Browser-timings

Lib for measure client browsers performance. Automatically collects performance data on page load.

## Installation

Install npm package

```bash
npm i --save @tinkoff/browser-timings
```

## How to

```ts
import { browserTimings } from '@tinkoff/browser-timings';

window.addEventListener('load', () => {
  setTimeout(() => {
    // setTimeout is necessary in order to get metrics about loadEventEnd
    const perfData = browserTimings();
  }, 0);
});
```

After executing `perfData` will contain client performance metrics which may be send to any external system for further analysis.

> Call of the library should be executed only after page load in order to get actual data. Otherwise, it may return empty object without data.

## Lib interface

```tsx
export interface Timings {
  /* Connection timing from client to server */
  connection: number;
  /* How much time backend was preparing response */
  backend: number;
  /* Page download to client */
  pageDownload: number;
  /* Timing of first paint for a page */
  'first-paint': number;
  /* Timing when DOM becomes interactive */
  domInteractive: number;
  /* DOM building is complete */
  domComplete: number;
  /* Page and every resource were loaded */
  pageLoadTime: number;
  /* Common information about resources and its loading time grouped by type */
  download: {
    html: TimingResource;
    js: TimingResource;
    css: TimingResource;
    img: TimingResource;
    font: TimingResource;
    other: TimingResource;
  };
}

interface TimingResource {
  /* Timing of resource loading */
  duration: number;
  /* Byte-size measure of data used by resource */
  encodedDecodeSize: number;
  /* Byte-size measure of data transferred by network. Calculating difference between encodedDecodeSize - transferSize may reveal how much data where stored in browser cache */
  transferSize: number;
}
```
