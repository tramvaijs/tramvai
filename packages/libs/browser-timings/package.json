{
  "name": "@tinkoff/browser-timings",
  "version": "0.0.0-stub",
  "initialVersion": "0.8.18",
  "description": "Lightweighted browser timings metrics exporter",
  "main": "./lib/browser-timings.js",
  "typings": "./lib/browser-timings.d.ts",
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git@github.com:Tinkoff/tramvai.git"
  },
  "scripts": {
    "build": "tramvai-build --forPublish --preserveModules",
    "watch": "tsc -w"
  },
  "sideEffects": false,
  "gitHead": "8e826a214c87b188fc4d254cdd8f2a2b2c55f3a8",
  "module": "./lib/browser-timings.es.js",
  "license": "Apache-2.0"
}
