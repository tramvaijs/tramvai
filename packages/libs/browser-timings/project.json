{
  "name": "@tinkoff/browser-timings",
  "sourceRoot": "packages/libs/browser-timings/src",
  "projectType": "library",
  "targets": {
    "build-publish": {
      "outputs": [
        "packages/libs/browser-timings/package.json",
        "packages/libs/browser-timings/lib"
      ],
      "executor": "@tramvai/nx-plugin:build"
    }
  }
}
