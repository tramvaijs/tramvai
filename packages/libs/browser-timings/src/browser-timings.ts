import type { TimingResource } from './browser-timings.h';
import { Timings } from './browser-timings.h';

const toSecond = (timeStamp: number) => Math.round(timeStamp) / 1000;

const testString = (pattern: RegExp) => (text: string) => pattern.test(text);
const isJs = testString(/\.js$/);
const isCSS = testString(/\.css$/);
const isImg = testString(/\.(?:svg|png|jpeg|webp|jpg)$/);
const isFont = testString(/\.(?:woff|eot|woff2|ttf)/);
enum resourceType {
  html = 'html',
  js = 'js',
  css = 'css',
  img = 'img',
  font = 'font',
  other = 'other',
}

const getType = (name: string) => {
  switch (true) {
    case isCSS(name):
      return resourceType.css;
    case isJs(name):
      return resourceType.js;
    case isImg(name):
      return resourceType.img;
    case isFont(name):
      return resourceType.font;
    default:
      return resourceType.other;
  }
};

const getEntriesByType = (performance: Performance, type: string) => {
  if (performance.getEntriesByType) {
    return performance.getEntriesByType(type);
  }
};

const getPaintMetric = (performance: Performance) => {
  const paint = getEntriesByType(performance, 'paint');

  if (paint && paint[0]) {
    return paint[0].startTime;
  }

  return 0;
};

const getTimingResourceValue = (tr: any, key: keyof TimingResource): number => {
  return (tr && tr[key]) || 0;
};

const calculateMetric = (prev: any, next: PerformanceResourceTiming): TimingResource => {
  return {
    duration: getTimingResourceValue(prev, 'duration') + toSecond(next.duration),
    transferSize: getTimingResourceValue(prev, 'transferSize') + next.transferSize,
    encodedBodySize: getTimingResourceValue(prev, 'encodedBodySize') + next.encodedBodySize,
  };
};

const getResourceMetrics = (performance: Performance) => {
  const result = {} as Record<string, any>;

  const entries = getEntriesByType(performance, 'resource') as PerformanceResourceTiming[];
  if (entries && entries.length > 0) {
    const countByType: Partial<Record<resourceType, number>> = {};

    entries.forEach((entry) => {
      const name = getType(entry.name.split('?', 1)[0]);

      if (resourceType[name]) {
        if (!countByType[name]) {
          countByType[name] = 0;
        }
        countByType[name]!++;

        result[name] = calculateMetric(result[name], entry);
        result[name].count = countByType[name];
      }
    });
  }

  const pageNav = getEntriesByType(performance, 'navigation') as PerformanceNavigationTiming[];
  if (pageNav && pageNav.length > 0) {
    result[resourceType.html] = calculateMetric({}, pageNav[0]);
  }

  return result;
};

type TimingsList = Array<{ key: string; start: number; end: number }>;

export function browserTimings(): Timings | {} {
  let performance;
  try {
    // Провряем, есть ли имплементация timing
    performance = window.performance;
    if (!performance.timing || !performance.timing.loadEventEnd) {
      // Если loadEventEnd равен 0, то библиотеку запустили раньше, чем нужно
      if (performance.timing.loadEventEnd === 0) {
        // eslint-disable-next-line no-console
        console.error('Library used at browser initialization page. Reed docs');
      }
      throw new Error('Not implemented timing api');
    }
  } catch (e) {
    return {};
  }

  const {
    requestStart,
    responseStart,
    responseEnd,
    domComplete,
    loadEventEnd,
    domInteractive,
    navigationStart,
  } = performance.timing;
  const timingCandidates: TimingsList = [
    {
      key: 'connection',
      start: navigationStart,
      end: requestStart,
    },
    {
      key: 'backend',
      start: requestStart,
      end: responseStart,
    },
    {
      key: 'pageDownload',
      start: responseStart,
      end: responseEnd,
    },
    {
      key: 'first-paint',
      start: 0,
      end: getPaintMetric(performance),
    },
    {
      key: 'domInteractive',
      start: navigationStart,
      end: domInteractive,
    },
    {
      key: 'domComplete',
      start: navigationStart,
      end: domComplete,
    },
    {
      key: 'pageLoadTime',
      start: navigationStart,
      end: loadEventEnd,
    },
  ];
  const timings: any = {};

  timingCandidates.forEach((el) => {
    const time = toSecond(el.end - el.start);
    // Отсеиваем не стандартные значения меньше 0 секунд и больше 33 минут
    timings[el.key] = time > 0 || time < 2000 ? time : undefined;
  });

  timings.download = getResourceMetrics(performance);

  return timings as Timings;
}

export { Timings };
