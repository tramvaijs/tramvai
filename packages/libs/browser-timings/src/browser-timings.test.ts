/**
 * @jest-environment jsdom
 */

import { browserTimings } from './browser-timings';

const paintMock = [
  {
    duration: 0,
    entryType: 'paint',
    name: 'first-paint',
    startTime: 1078.9400000066962,
  },
];

const timingMock = {
  navigationStart: 1543996403132,
  unloadEventStart: 0,
  unloadEventEnd: 0,
  redirectStart: 0,
  redirectEnd: 0,
  fetchStart: 1543996403146,
  domainLookupStart: 1543996403148,
  domainLookupEnd: 1543996403148,
  connectStart: 1543996403148,
  connectEnd: 1543996403153,
  secureConnectionStart: 1543996403149,
  requestStart: 1543996403155,
  responseStart: 1543996403399,
  responseEnd: 1543996403413,
  domLoading: 1543996403410,
  domInteractive: 1543996403862,
  domContentLoadedEventStart: 1543996404120,
  domContentLoadedEventEnd: 1543996404127,
  domComplete: 1543996404794,
  loadEventStart: 1543996404794,
  loadEventEnd: 1543996404795,
};

const navigateMock = [
  {
    name: 'https://tinkoff.ru/',
    entryType: 'navigation',
    startTime: 0,
    duration: 1890.1200000109384,
    initiatorType: 'navigation',
    nextHopProtocol: 'http/2+quic/44',
    workerStart: 0,
    redirectStart: 0,
    redirectEnd: 0,
    fetchStart: 2.885000008973293,
    domainLookupStart: 9.734999999636784,
    domainLookupEnd: 9.745000003022142,
    connectStart: 9.745000003022142,
    connectEnd: 34.25000001152512,
    secureConnectionStart: 0,
    requestStart: 14.290000006440096,
    responseStart: 807.7250000060303,
    responseEnd: 989.5749999996042,
    transferSize: 87483,
    encodedBodySize: 87128,
    decodedBodySize: 350459,
    serverTiming: [],
    unloadEventStart: 816.7250000115018,
    unloadEventEnd: 817.0350000000326,
    domInteractive: 1347.780000010971,
    domContentLoadedEventStart: 1347.829999998794,
    domContentLoadedEventEnd: 1488.2100000104401,
    domComplete: 1886.0350000031758,
    loadEventStart: 1886.0600000043632,
    loadEventEnd: 1890.1200000109384,
    type: 'reload',
    redirectCount: 0,
  },
];

const resourceMock = [
  {
    connectEnd: 1011.3400000118418,
    connectStart: 1011.3400000118418,
    decodedBodySize: 1211545,
    domainLookupEnd: 1011.3400000118418,
    domainLookupStart: 1011.3400000118418,
    duration: 0,
    encodedBodySize: 239719,
    entryType: 'resource',
    fetchStart: 1011.3400000118418,
    initiatorType: 'script',
    name: 'https://tinkoff.ru/script_foot.js',
    nextHopProtocol: 'http/2+quic/44',
    redirectEnd: 0,
    redirectStart: 0,
    requestStart: 1011.3400000118418,
    responseEnd: 1011.3400000118418,
    responseStart: 1011.3400000118418,
    secureConnectionStart: 1011.3400000118418,
    serverTiming: [],
    startTime: 1011.3400000118418,
    transferSize: 0,
    workerStart: 0,
  },
  {
    connectEnd: 963.7900000088848,
    connectStart: 963.7900000088848,
    decodedBodySize: 204267,
    domainLookupEnd: 963.7900000088848,
    domainLookupStart: 963.7900000088848,
    duration: 5.095000000437722,
    encodedBodySize: 38422,
    entryType: 'resource',
    fetchStart: 963.7900000088848,
    initiatorType: 'link',
    name: 'https://tinkoff.ru/devsite.css',
    nextHopProtocol: 'http/2+quic/44',
    redirectEnd: 0,
    redirectStart: 0,
    requestStart: 965.585000012652,
    responseEnd: 968.8850000093225,
    responseStart: 967.3450000118464,
    secureConnectionStart: 0,
    serverTiming: [],
    startTime: 963.7900000088848,
    transferSize: 0,
    workerStart: 0,
  },
];

describe('browser-timings', () => {
  it('performance объект пустой', () => {
    const result = browserTimings();

    expect(result).toMatchSnapshot();
  });

  it('Получение перфоманс метрик', () => {
    Object.defineProperty(window, 'performance', {
      value: {
        timeOrigin: 1543996403132.6418,
        timing: timingMock,
        navigation: { type: 0, redirectCount: 0 },
        getEntriesByType: (type) => {
          switch (type) {
            case 'navigation':
              return navigateMock;
            case 'resource':
              return resourceMock;
            case 'paint':
              return paintMock;
            default:
              return [];
          }
        },
      },
    });
    const result = browserTimings();

    expect(result).toMatchSnapshot();
  });
});
