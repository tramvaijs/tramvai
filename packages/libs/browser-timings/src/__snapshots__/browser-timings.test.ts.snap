// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browser-timings performance объект пустой 1`] = `{}`;

exports[`browser-timings Получение перфоманс метрик 1`] = `
{
  "backend": 0.244,
  "connection": 0.023,
  "domComplete": 1.662,
  "domInteractive": 0.73,
  "download": {
    "css": {
      "count": 1,
      "duration": 0.005,
      "encodedBodySize": 38422,
      "transferSize": 0,
    },
    "html": {
      "duration": 1.89,
      "encodedBodySize": 87128,
      "transferSize": 87483,
    },
    "js": {
      "count": 1,
      "duration": 0,
      "encodedBodySize": 239719,
      "transferSize": 0,
    },
  },
  "first-paint": 1.079,
  "pageDownload": 0.014,
  "pageLoadTime": 1.663,
}
`;
